### PR TITLE
document thoughts on code generation

### PR DESCRIPTION
**Pull request by jaypipes**
Adds a document discussing options for code generation and a diagram
with how a multi-phase hybrid custom+controller-runtime generator might
work.

Issue #4

### CODE REVIEW

File: docs/code-generation.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+To generate the CRDs and controller stub code, we investigated a number of
+options:
+
+* home-grown custom code generator
+* [kudo](https://kudo.dev)
+* [kubebuilder](github.com/kubernetes-sigs/kubebuilder)
+* a hybrid custom code generator + sigs.kubernetes.io/controller-tools (CR)
+
+The original AWS Service Operator used a [custom-built generator][1] that
+processed [YAML files][2] describing the AWS service and used [templates][3] to
```
Comment (mhausenblas) [inline docs/code-generation.md]: s/YAML files/YAML manifests

File: docs/code-generation.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+[generate CRDs][4], the [controller code][5] itself and the [Go types][6] that
+mirror the CRDs. It's worth noting that the CRDs *and* the controller code that
```
Comment (mhausenblas) [inline docs/code-generation.md]: s/mirror the CRDs/represent the CRDs in memory

File: docs/code-generation.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+[3]: https://github.com/amazon-archives/aws-service-operator/tree/master/code-generation/pkg/codegen/assets
+[4]: https://github.com/amazon-archives/aws-service-operator/blob/b4befd62322a57ac78aa39ea08771fc32912592a/code-generation/pkg/codegen/assets/aws-service-operator.yaml.templ#L13-L31
+[5]: https://github.com/amazon-archives/aws-service-operator/blob/master/code-generation/pkg/codegen/assets/operator.go.templ
+[6]: https://github.com/amazon-archives/aws-service-operator/blob/master/code-generation/pkg/codegen/assets/types.go.templ
+
+kudo is a platform for building Kubernetes Operators. It stores state in its
+own kudo.dev CRDs and allows users to define "plans" for a deployed application
+to deploy itself. We determined that kudo was not a particularly good fit for
+ASO for a couple reasons. First, we needed a way to generate CRDs in several
+API groups (s3.aws.com and iam.aws.com for example) and the ASO controller code
```
Comment (mhausenblas) [inline docs/code-generation.md]: Given that we own `service-operator.aws` I suggest we use it for FQDNs and CR{D} related examples?

File: docs/code-generation.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+isn't deploying an "application" that needs to have a controlled deployment
+plan. Instead, ASO is a controller that facilitates creation and management of
+various AWS service objects using Kubernetes CRD instances.
+
+kubebuilder is the the recommended upstream tool for generating CRDs and
```
Comment (mhausenblas) [inline docs/code-generation.md]: Mickey Blue Eyes disapproves :)

s/kubebuilder is the the/kubebuilder is the

File: docs/code-generation.md [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+controller stub code. It is a Golang binary that creates the scaffolding for
```
Comment (mhausenblas) [inline docs/code-generation.md]: Nit: While `Golang` is not incorrect, I'd prefer the simpler, short, and in our context unambiguous `Go`.

File: docs/code-generation.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+[7]: https://github.com/kubernetes-sigs/kubebuilder/issues/1268
+
+## Hybrid custom+controller-runtime proposal
+
+![Multi-phase approach to code generation for ASO](images/multi-phase-code-generation.png)
+
+This approach uses multiple phases of code generation.
+
+The first code generation phase consumes model information from a canonical
```
Comment (mhausenblas) [inline docs/code-generation.md]: I'm not entirely sure, hence a clarification question: are you talking about using Go pragma, esp. the `//go:generate` directive for [generating](https://blog.golang.org/generate) this?

File: docs/code-generation.md [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+Next, we will need to generate some skeleton code for the reconciling
+controller handling the new API along with the YAML files representing the CRDs
+that will get loaded into kube-apiserver and the YAML files for RBAC and
+OpenAPI v3 schemas for the CRDs. This is where the [`controller-gen`][9] tool
+from the sigs.kubernetes.io/controller-tools project will come in handy.
+
+[9]: https://github.com/kubernetes-sigs/controller-tools/blob/a5fa7b956b85a6e792bc7086fedf7107d62452b1/cmd/controller-gen/main.go
+
+We also want to generate some Golang code that "installs" a new reconciling
```
Comment (mhausenblas) [inline docs/code-generation.md]: I think we should separate the build and operate (install/upgrade/ALM) concerns.

File: docs/code-generation.md (hunk 1 above)
Comment (jaypipes) [reply docs/code-generation.md]: done.

File: docs/code-generation.md (hunk 2 above)
Comment (jaypipes) [reply docs/code-generation.md]: done.

File: docs/code-generation.md (hunk 3 above)
Comment (jaypipes) [reply docs/code-generation.md]: as chatted about on Slack call, yeah, I will change this to `s3.service-operator.aws`

File: docs/code-generation.md (hunk 4 above)
Comment (jaypipes) [reply docs/code-generation.md]: done.

File: docs/code-generation.md (hunk 5 above)
Comment (jaypipes) [reply docs/code-generation.md]: done :)

File: docs/code-generation.md (hunk 6 above)
Comment (jaypipes) [reply docs/code-generation.md]: Yes, the first phase of code generation can be a Go file that reads the model files from `aws-sdk-go/models` and creates the Go type files that are used as input for the second phase of code generation.

File: docs/code-generation.md (hunk 7 above)
Comment (jaypipes) [reply docs/code-generation.md]: done.

**Pull request merged**